File: celextrix_impossible_return_v3/verifier.mjs
#!/usr/bin/env node
import fs from "node:fs";
import {
  PROTOCOL, STAGE_A_POINTS, STAGE_B_POINTS, MIN_ROUTE_LENGTH, MAX_ROUTE_LENGTH,
  mod, canonicalJson,
} from "./protocol.mjs";

function same(a,b){ return a.length===b.length && a.every((v,i)=>Array.isArray(v)?same(v,b[i]):v===b[i]); }
function toBig(v){ if(!Number.isInteger(v) || v<0) throw new Error("PROOF_INTEGER_INVALID"); return BigInt(v); }
function vec(v,n){ if(!Array.isArray(v)||v.length!==n) throw new Error("PROOF_VECTOR_SHAPE"); return v.map(toBig); }
function matrix(v,r,c){ if(!Array.isArray(v)||v.length!==r) throw new Error("PROOF_MATRIX_SHAPE"); return v.map(row=>vec(row,c)); }
function step(ch){ if(ch==="N")return[0n,1n]; if(ch==="E")return[1n,0n]; if(ch==="S")return[0n,-1n]; if(ch==="W")return[-1n,0n]; throw new Error("ROUTE_SYMBOL"); }
function routeState(route){
  if(typeof route!=="string"||!/^[NESW]+$/.test(route)) throw new Error("ROUTE_GRAMMAR");
  if(route.length<MIN_ROUTE_LENGTH||route.length>MAX_ROUTE_LENGTH) throw new Error("ROUTE_LENGTH");
  let x=0n,y=0n; const t=[0n,0n,0n,0n];
  for(const ch of route){ const [dx,dy]=step(ch); const nx=x+dx,ny=y+dy; const w=x*ny-y*nx; const sx=x+nx,sy=y+ny; const e=[w,w*sx,w*sy,w*(sx*sx+sy*sy)]; for(let i=0;i<4;i++)t[i]=mod(t[i]+e[i]); x=nx;y=ny; }
  if(x!==0n||y!==0n) throw new Error("ROUTE_NOT_CLOSED");
  return t;
}
function V(points){ return points.map(z=>[1n,z,mod(z*z),mod(z*z*z)]); }
function mv(m,v){ return m.map(row=>mod(row.reduce((s,a,i)=>s+a*v[i],0n))); }
function mm(a,b){ return a.map(row=>b[0].map((_,j)=>mod(row.reduce((s,x,i)=>s+x*b[i][j],0n)))); }
function I(n){ return Array.from({length:n},(_,r)=>Array.from({length:n},(_,c)=>r===c?1n:0n)); }

export function verifyProofObject(obj){
  if(!obj||typeof obj!=="object"||Array.isArray(obj)) throw new Error("PROOF_NOT_OBJECT");
  if(obj.protocol!==PROTOCOL) throw new Error("PROTOCOL_MISMATCH");
  if(obj.field_prime!==101) throw new Error("FIELD_MISMATCH");
  if(obj.stage!=="A"&&obj.stage!=="B") throw new Error("STAGE_INVALID");
  const ap=vec(obj.stage_a_points,3), bp=vec(obj.stage_b_points,4);
  if(!same(ap,STAGE_A_POINTS)||!same(bp,STAGE_B_POINTS)) throw new Error("OBSERVER_POINTS_MISMATCH");
  const k=vec(obj.stage_a_kernel_witness,4); if(k.every(x=>x===0n)||!same(mv(V(ap),k),[0n,0n,0n])) throw new Error("STAGE_A_KERNEL_WITNESS_INVALID");
  const inv=matrix(obj.stage_b_inverse_witness,4,4), vb=V(bp); if(!same(mm(inv,vb),I(4))||!same(mm(vb,inv),I(4))) throw new Error("STAGE_B_INVERSE_WITNESS_INVALID");
  const L=obj.left,R=obj.right; if(!L||!R||L.route===R.route) throw new Error("PATHS_NOT_DISTINCT");
  const lt=routeState(L.route), rt=routeState(R.route), ldecl=vec(L.target,4), rdecl=vec(R.target,4);
  if(!same(lt,ldecl)||!same(rt,rdecl)) throw new Error("TARGET_TRANSCRIPT_INVALID");
  const points=obj.stage==="A"?ap:bp, lo=mv(V(points),lt), ro=mv(V(points),rt), lodecl=vec(L.observer,points.length), rodecl=vec(R.observer,points.length);
  if(!same(lo,lodecl)||!same(ro,rodecl)) throw new Error("OBSERVER_TRANSCRIPT_INVALID");
  if(obj.stage==="B" && (!same(mv(inv,lo),lt)||!same(mv(inv,ro),rt))) throw new Error("INVERSE_RECONSTRUCTION_FAILURE");
  const observerEquivalent=same(lo,ro), targetDistinct=!same(lt,rt);
  let result="NO_BREAK";
  if(obj.stage==="A"&&observerEquivalent&&targetDistinct) result="BREAK_ACCEPTED_STAGE_A_CONTROL";
  if(obj.stage==="B"&&observerEquivalent&&targetDistinct) result="THEOREM_CONTRADICTION_STAGE_B";
  return { protocol:PROTOCOL, result, stage:obj.stage, theorem_witness_closed:true, transcript_closed:true, observer_equivalent:observerEquivalent, target_distinct:targetDistinct };
}

export function verifyProofRaw(raw){
  const obj=JSON.parse(raw); const canonical=canonicalJson(obj)+"\n"; if(canonical!==raw) throw new Error("PROOF_NOT_CANONICAL"); return verifyProofObject(obj);
}

if(process.argv[1]&&process.argv[1].endsWith("verifier.mjs")){
  try{ const f=process.argv[2]; const raw=f&&f!=="-"?fs.readFileSync(f,"utf8"):fs.readFileSync(0,"utf8"); const out=verifyProofRaw(raw); process.stdout.write(canonicalJson(out)+"\n"); if(out.result==="THEOREM_CONTRADICTION_STAGE_B")process.exitCode=3; }
  catch(e){ process.stdout.write(canonicalJson({protocol:PROTOCOL,result:"PROOF_REJECTED",error:e instanceof Error?e.message:String(e)})+"\n"); process.exitCode=2; }
}

File: celextrix_impossible_return_v3/verifier.py
#!/usr/bin/env python3
import json, sys
PROTOCOL="celextrix-impossible-return-math-liberation-v3"; P=101
AP=[1,2,3]; BP=[1,2,3,4]; MINLEN=8; MAXLEN=64

def mod(x): return x%P

def canon(x):
    if x is None or isinstance(x,(bool,int,float,str)): return json.dumps(x,separators=(',',':'),ensure_ascii=True)
    if isinstance(x,list): return '['+','.join(canon(v) for v in x)+']'
    if isinstance(x,dict): return '{'+','.join(json.dumps(k)+':'+canon(x[k]) for k in sorted(x))+'}'
    raise TypeError()

def vec(v,n):
    if not isinstance(v,list) or len(v)!=n or any(type(x) is not int or x<0 for x in v): raise ValueError('PROOF_VECTOR_SHAPE')
    return v

def matrix(v,r,c):
    if not isinstance(v,list) or len(v)!=r: raise ValueError('PROOF_MATRIX_SHAPE')
    return [vec(row,c) for row in v]

def route_target(route):
    if not isinstance(route,str) or any(ch not in 'NESW' for ch in route): raise ValueError('ROUTE_GRAMMAR')
    if not MINLEN<=len(route)<=MAXLEN: raise ValueError('ROUTE_LENGTH')
    x=y=0; t=[0,0,0,0]; d={'N':(0,1),'E':(1,0),'S':(0,-1),'W':(-1,0)}
    pts=[(0,0)]
    for ch in route:
        dx,dy=d[ch]; x+=dx; y+=dy; pts.append((x,y))
    if (x,y)!=(0,0): raise ValueError('ROUTE_NOT_CLOSED')
    for (x0,y0),(x1,y1) in zip(pts,pts[1:]):
        w=x0*y1-y0*x1; sx=x0+x1; sy=y0+y1
        edge=[w,w*sx,w*sy,w*(sx*sx+sy*sy)]
        t=[mod(a+b) for a,b in zip(t,edge)]
    return t

def V(points): return [[1,z,mod(z*z),mod(z*z*z)] for z in points]
def mv(m,v): return [mod(sum(a*b for a,b in zip(row,v))) for row in m]
def mm(a,b): return [[mod(sum(row[i]*b[i][j] for i in range(len(b)))) for j in range(len(b[0]))] for row in a]
def ident(n): return [[1 if i==j else 0 for j in range(n)] for i in range(n)]

def verify(obj):
    if not isinstance(obj,dict): raise ValueError('PROOF_NOT_OBJECT')
    if obj.get('protocol')!=PROTOCOL: raise ValueError('PROTOCOL_MISMATCH')
    if obj.get('field_prime')!=101: raise ValueError('FIELD_MISMATCH')
    stage=obj.get('stage')
    if stage not in ('A','B'): raise ValueError('STAGE_INVALID')
    ap=vec(obj.get('stage_a_points'),3); bp=vec(obj.get('stage_b_points'),4)
    if ap!=AP or bp!=BP: raise ValueError('OBSERVER_POINTS_MISMATCH')
    k=vec(obj.get('stage_a_kernel_witness'),4)
    if all(x==0 for x in k) or mv(V(ap),k)!=[0,0,0]: raise ValueError('STAGE_A_KERNEL_WITNESS_INVALID')
    inv=matrix(obj.get('stage_b_inverse_witness'),4,4); vb=V(bp)
    if mm(inv,vb)!=ident(4) or mm(vb,inv)!=ident(4): raise ValueError('STAGE_B_INVERSE_WITNESS_INVALID')
    L,R=obj.get('left'),obj.get('right')
    if not isinstance(L,dict) or not isinstance(R,dict) or L.get('route')==R.get('route'): raise ValueError('PATHS_NOT_DISTINCT')
    lt=route_target(L.get('route')); rt=route_target(R.get('route'))
    if lt!=vec(L.get('target'),4) or rt!=vec(R.get('target'),4): raise ValueError('TARGET_TRANSCRIPT_INVALID')
    points=ap if stage=='A' else bp; lo=mv(V(points),lt); ro=mv(V(points),rt)
    if lo!=vec(L.get('observer'),len(points)) or ro!=vec(R.get('observer'),len(points)): raise ValueError('OBSERVER_TRANSCRIPT_INVALID')
    if stage=='B' and (mv(inv,lo)!=lt or mv(inv,ro)!=rt): raise ValueError('INVERSE_RECONSTRUCTION_FAILURE')
    oe=lo==ro; td=lt!=rt; result='NO_BREAK'
    if stage=='A' and oe and td: result='BREAK_ACCEPTED_STAGE_A_CONTROL'
    if stage=='B' and oe and td: result='THEOREM_CONTRADICTION_STAGE_B'
    return {'protocol':PROTOCOL,'result':result,'stage':stage,'theorem_witness_closed':True,'transcript_closed':True,'observer_equivalent':oe,'target_distinct':td}

def main():
    raw=open(sys.argv[1],encoding='utf8').read() if len(sys.argv)>1 and sys.argv[1]!='-' else sys.stdin.read()
    try:
        obj=json.loads(raw)
        if canon(obj)+'\n'!=raw: raise ValueError('PROOF_NOT_CANONICAL')
        out=verify(obj); print(canon(out)); sys.exit(3 if out['result']=='THEOREM_CONTRADICTION_STAGE_B' else 0)
    except Exception as e:
        print(canon({'protocol':PROTOCOL,'result':'PROOF_REJECTED','error':str(e)})); sys.exit(2)
if __name__=='__main__': main()
